docs/SPECIFICATION.md

```diff
@@ -424,6 +427,49 @@ A LinkRemove in a message `m` is valid only if it passes these validations:
 
 1.  `m.data.type` must be `MESSAGE_TYPE_LINK_REMOVE`
 
+## 1.7 Username Proof
+
+```protobuf
+enum UserNameType {
+  USERNAME_TYPE_NONE = 0;
+  USERNAME_TYPE_ENS_FNAME = 1;
+  USERNAME_TYPE_ENS_L1 = 2;
+}
+
+message UserNameProofBody {
+  uint64 timestamp = 1;
+  bytes name = 2;
+  bytes owner = 3;
+  bytes signature = 4;
+  uint64 fid = 5;
+  UserNameType type = 6;
+}
+```
+
+A UsernameProof message `m` must pass these validations:
+
+1. `m.signature_scheme` must be `SIGNATURE_SCHEME_ED25519`.
+2. `m.data.body` must be `UserNameProofBody`.
+3. `m.data.body.timestamp` must be ≤ 10 mins ahead of current timestamp.
+4. `m.data.body.fid` must be a known fid.
+
+A UsernameProof message `m` of type `USERNAME_TYPE_ENS_FNAME` must also pass these validations:
+
+1. `m.data.body.name` name must match the regular expression `/^[a-z0-9][a-z0-9-]{0,15}$/`.
+2. `m.data.body.owner` must be the custody address of the fid.
+3. `m.data.body.signature` must be a valid ECDSA signature on the EIP-712 Username Proof message from the owner or the public key of the fname server.
+
+A UsernameProof message `m` of type `USERNAME_TYPE_ENS_L1` must also pass these validations:
+
+1. `m.data.body.name` name must:
+   1. be a valid, unexpired ENS name
+   2. match the regular expression `/^[a-z0-9][a-z0-9-]{0,15}$/`
+   3. end in `.eth`
+2. `m.data.body.owner` must:
+   1. be an address that the user has a valid VerificationMessage for.
+   2. be the owner of the ENS name.
```

Comment:
ENS name must resolve to the `owner` is more accurate. The actual registrant or controller can be different.